core/src/main/java/com/segment/analytics/kotlin/core/Telemetry.kt

```diff
@@ -0,0 +1,272 @@
+
+    var httpClient: HTTPClient
+        get() = _httpClient
+        set(value) {
+            _httpClient = value
+        }
+
+    private val queue = mutableListOf<RemoteMetric>()
+    private var queueBytes = 0
+    private var queueSizeExceeded = false
+    private val seenErrors = mutableMapOf<String, Int>()
+    private var started = false
+
+    init {
+        CoroutineScope(Dispatchers.Default).launch {
+            while (isActive) {
+                if (!enable) return@launch
+                try {
+                    flush()
+                } catch (e: Throwable) {
+                    logError(e)
+                }
+                try {
+                    delay(flushTimer.toLong())
+                } catch (e: CancellationException) {
+                    // last flush before shutdown
+                    flush()
+                }
+
+            }
+        }
+    }
+
+    fun start() {
+        if (started) return
+        // Assume sampleRate is now set and everything in the queue hasn't had it applied
+        if (Math.random() > sampleRate) {
+            resetQueue()
+        } else {
+            queue.forEach {
+                it.value = (it.value / sampleRate).roundToInt()
+            }
+        }
+
+        started = true
+    }
+
+    fun increment(metric: String, tags: Map<String, String>) {
+        if (!enable) return
+        if (!metric.startsWith("analytics_mobile.")) return
+        if (tags.isEmpty()) return
+        if (Math.random() > sampleRate) return
+        if (queue.size >= maxQueueSize) return
+
+        addRemoteMetric(metric, tags, value = (1.0 / sampleRate).roundToInt())
+    }
+
+    fun error(metric:String, tags: Map<String, String>, log: String) {
+        if (!enable) return
+        if (!metric.startsWith("analytics_mobile.")) return
+        if (tags.isEmpty()) return
+        if (queue.size >= maxQueueSize) return
+
+        var filteredTags = tags
+        if (!sendWriteKeyOnError) filteredTags = tags.filterKeys { it != "writekey" }
+        var logData: String? = null
+        if (sendErrorLogData) logData = log
+
+        val errorKey = tags["error"]
+        if (errorKey != null) {
+            if (seenErrors.containsKey(errorKey)) {
+                seenErrors[errorKey] = seenErrors[errorKey]!! + 1
+                if (Math.random() > sampleRate) return
+                // Send how many we've seen after the first since we know for sure.
+                addRemoteMetric(metric, filteredTags, log=logData, value = seenErrors[errorKey]!!)
+                seenErrors[errorKey] = 0
+            } else {
+                addRemoteMetric(metric, filteredTags, log=logData)
+                flush()
+                seenErrors[errorKey] = 0 // Zero because it's already been sent.
+            }
+        }
+    }
+
+    fun flush() {
+        if (!started || !enable || queue.isEmpty()) return
+
+        if (_rateLimitTimer > 0 && _rateLimitTimer > (System.currentTimeMillis() / 1000).toInt()) {
```

Comment:
I think the first check (`> 0`) is redundant?

Reply:
...  Yes.  Yes it is.  Let's not talk about the 5 minutes I spent staring at this earlier today reassuring myself it made sense.